File: _includes/code/algorithms/99-problems/ocaml/linear_search.ml
open OUnit2

let rec linear_search arr t = match arr with 
  | [] -> false 
  | x::_ when x = t -> true 
  | _::xs -> linear_search xs t


let tests = "test suite for linear_search" >::: [
  "[] 10"   >:: (fun _ -> assert_equal false (linear_search [] 10));
  "[1] 1"   >:: (fun _ -> assert_equal true (linear_search [1] 1));
  "[1;2] 3" >:: (fun _ -> assert_equal false (linear_search [1; 2] 3));
  "[1;2] 4" >:: (fun _ -> assert_equal false (linear_search [1; 2] 4));
]

let run_tests = run_test_tt_main tests

File: _includes/code/algorithms/99-problems/rust/src/linear_search.rs
use std::cmp::PartialEq;

#[allow(dead_code)]
fn linear_search<T: PartialEq>(arr: Vec<T>, target: T) -> Option<usize> {
    for (index, value) in arr.into_iter().enumerate() {
        if value == target {
            return Some(index);
        }
    }

    None
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_match() {
        assert_eq!(linear_search::<i32>(vec![1, 2, 3, 4], 4), Some(3));
    }

    #[test]
    fn test_none() {
        assert_eq!(linear_search::<i32>(vec![1, 2, 3, 4], 6), None);
    }

    #[test]
    fn test_empty() {
        assert_eq!(linear_search::<i32>(vec![], 4), None);
    }
}
